File: _includes/code/howto/manage-data.delete.ts
// How-to: Manage data -> Delete objects - TypeScript examples
// run with: node --loader=ts-node/esm manage-data.delete.ts
import assert from 'assert';

// ================================
// ===== INSTANTIATION-COMMON =====
// ================================
import weaviate, { WeaviateClient } from 'weaviate-client';

const client: WeaviateClient = await weaviate.connectToWCD(
  process.env.WCD_URL,
 {
   authCredentials: new weaviate.ApiKey(process.env.WCD_API_KEY),
   headers: {
     'X-OpenAI-Api-Key': process.env.OPENAI_APIKEY,  // Replace with your inference API key
   }
 } 
)

// START DeleteObject // START DryRun // START DeleteBatch // START DeleteByIDBatch // START DeleteContain
const myCollection = client.collections.get('EphemeralObject')

// END DeleteObject // END DryRun // END DeleteBatch // END DeleteByIDBatch // END DeleteContain


// =========================
// ===== Delete object =====
// =========================

// START DeleteObject
let idToDelete = '...';  // replace with the id of the object you want to delete
// END DeleteObject
idToDelete = await myCollection.data.insert({
  name: 'EphemeralObjectA'
})

// START DeleteObject
// highlight-start
await myCollection.data.deleteById(idToDelete)
// highlight-end
// END DeleteObject

// Test
try {
  result = await myCollection.query.fetchObjectById(idToDelete)
  assert.equal(result, null);  // execution should not reach this point
} catch (e) {
  // This 404 error is EXPECTED, because the object was deleted
  // TODO: this behavior is inconsistent with the Python client, which returns None
  assert(e.message.includes(404));  // TODO: this should be a proper code - https://github.com/weaviate/weaviate/issues/2708#issuecomment-1582430931
}


// ==========================
// ===== Error handling =====
// ==========================

// START DeleteError
try {
  await myCollection.data.deleteById(idToDelete)
  // Returns undefined on success
} catch (e) {
  // 404 error if the id was not found
  console.error(e);
  // END DeleteError
  assert(e.message.includes('404'));
  // START DeleteError
}
// END DeleteError


// ===================
// ===== Dry run =====
// ===================
const N = 5;
for (let i = 1; i <= N; i++)
  await myCollection.data.insert({
    name: 'EphemeralObjectA' + i.toString(),
  })

{
// START DryRun
const response = await myCollection.data.deleteMany(
  myCollection.filter.byProperty('name').like('EphemeralObject'),
  // highlight-start
  {
    dryRun: true,
    verbose: true
  }
  // highlight-end
)

console.log(response);
// END DryRun
}


// ========================
// ===== Batch delete =====
// ========================
{
// START DeleteBatch
// highlight-start
const response = await myCollection.data.deleteMany(
  myCollection.filter.byProperty('name').like('EphemeralObject*')
)
// highlight-end

console.log(JSON.stringify(response))
// END DeleteBatch

// Test
assert.equal(response.matches, 0);
// assert.equal(response.matches, N); //fix

const leftovers = await myCollection.data.deleteMany(
  myCollection.filter.byProperty('name').like('EphemeralObject*')
)

assert.equal(leftovers?.matches, N);
}
// =====================================
// ===== Batch delete with Contain =====
// =====================================


// START DeleteContain
// highlight-start
await myCollection.data.deleteMany(
  myCollection.filter.byProperty('name').containsAny(['europe', 'asia'])
)
// highlight-end
// END DeleteContain

// ==============================
// ===== Batch delete by ID =====
// ==============================
{
// START DeleteByIDBatch
const response = await myCollection.query.fetchObjects({limit: 3})

const idList = response.objects.map(o => o.uuid)

// highlight-start
await myCollection.data.deleteMany(
  myCollection.filter.byId().containsAny(idList)
)
// highlight-end
// END DeleteByIDBatch
}

File: _includes/code/howto/manage-data.delete.py
# How-to: Manage-data -> Update objects - Python examples
import os

# ================================
# ===== INSTANTIATION-COMMON =====
# ================================

import weaviate

client = weaviate.connect_to_local()

try:
    client.collections.delete("EphemeralObject")
    client.collections.create("EphemeralObject")

    # =========================
    # ===== Delete object =====
    # =========================

    # START DeleteObject
    uuid_to_delete = "..."  # replace with the id of the object you want to delete
    # END DeleteObject

    collection = client.collections.get("EphemeralObject")
    uuid_to_delete = collection.data.insert({
        "name": "EphemeralObjectA",
    })

    # Test insertion
    assert collection.query.fetch_object_by_id(uuid_to_delete)  # Should not fail if object exists

    # START DeleteObject

    collection = client.collections.get("EphemeralObject")
    collection.data.delete_by_id(
        uuid_to_delete
    )
    # END DeleteObject

    # Test
    response = collection.query.fetch_object_by_id(uuid_to_delete)  # Should return None
    assert response == None


    # # ==========================
    # # ===== Error handling =====
    # # ==========================

    # # START DeleteError
    # try:
    #     client.data_object.delete(
    #         uuid=uuid_to_delete,
    #         collection_name="EphemeralObject",
    #     )
    #     # Returns None on success
    # except weaviate.exceptions.UnexpectedStatusCodeException as e:
    #     # 404 error if the id was not found
    #     print(e)
    # # END DeleteError
    #     # Test
    #     assert e.status_code == 404


    # ========================
    # ===== Batch delete =====
    # ========================
    N = 5
    for i in range(N):
        collection.data.insert({
            "name": f"EphemeralObject_{i}",
        })

    # Test insertion
    response = collection.aggregate.over_all(total_count=True)
    assert response.total_count == 5

    # START DeleteBatch
    import weaviate.classes as wvc

    collection = client.collections.get("EphemeralObject")
    collection.data.delete_many(
        # highlight-start
        where=wvc.query.Filter.by_property("name").like("EphemeralObject*")
        # highlight-end
    )
    # END DeleteBatch

    # Test deletion
    response = collection.aggregate.over_all(total_count=True)
    assert response.total_count == 0


    # ============================
    # ====== Delete Contains =====
    # ============================

    collection.data.insert_many([{
        "name": "asia",
    },{
        "name": "europe"
    }])

    # START DeleteContains
    import weaviate.classes as wvc

    collection = client.collections.get("EphemeralObject")
    collection.data.delete_many(
        # highlight-start
        where=wvc.query.Filter.by_property("name").contains_any(["europe", "asia"])
        # highlight-end
    )
    # END DeleteContains


    # ===================
    # ===== Dry run =====
    # ===================

    client.collections.delete("EphemeralObject")
    client.collections.create("EphemeralObject")

    collection = client.collections.get("EphemeralObject")

    N = 5
    for i in range(N):
        collection.data.insert({
            "name": f"EphemeralObject_{i}",
        })

    # START DryRun
    import weaviate.classes as wvc

    collection = client.collections.get("EphemeralObject")
    result = collection.data.delete_many(
        where=wvc.query.Filter.by_property("name").like("EphemeralObject*"),
        # highlight-start
        dry_run=True,
        verbose=True
        # highlight-end
    )

    print(result)
    # END DryRun

    assert result.matches == N
    assert collection.aggregate.over_all(total_count=True).total_count == N

    # =================================
    # ===== Batch delete with IDs =====
    # =================================
    client.collections.delete("EphemeralObject")
    client.collections.create("EphemeralObject")

    N = 5
    for i in range(N):
        collection.data.insert({
            "name": f"EphemeralObject_{i}",
        })

    # Test insertion
    response = collection.aggregate.over_all(total_count=True)
    assert response.total_count == 5

    collection = client.collections.get("EphemeralObject")
    response = collection.query.fetch_objects(limit=3)
    ids = [o.uuid for o in response.objects]

    # START DeleteByIDBatch
    import weaviate.classes as wvc

    collection = client.collections.get("EphemeralObject")

    response = collection.query.fetch_objects(limit=3)  # Fetch 3 object IDs
    ids = [o.uuid for o in response.objects]  # These can be lists of strings, or `UUID` objects

    collection.data.delete_many(
        # highlight-start
        where=wvc.query.Filter.by_id().contains_any(ids)  # Delete the 3 objects
        # highlight-end
    )
    # END DeleteByIDBatch

    # Test deletion
    response = collection.aggregate.over_all(total_count=True)
    assert response.total_count == 2

finally:
    client.close()
